Move prebuilds to GitHub Actions

zbjornson opened the pull request:
@chearon this basically copies your `ci/` folder into `prebuild/`, and ports the travis config + `install.sh` to GitHub actions. I haven't started Windows yet (and I've never used msys, but was reading through https://github.com/actions/virtual-environments/issues/30 and it looks definitely possible).

[There's no manual trigger for GitHub actions yet](https://github.community/t5/GitHub-Actions/GitHub-Actions-Manual-Trigger-Approvals/td-p/31504), but hopefully builds are reliable enough that doing this on release creation, or at least push, is viable.

zbjornson commented:
All three platforms working!

I set this to basically be manually triggered (pushes to master that change `.github/workflows/prebuild.yaml`) so we have more control over it, at least initially.

zbjornson commented on prebuild/Windows/preinstall.sh:
```diff
@@ -0,0 +1,38 @@
+  mingw64/mingw-w64-x86_64-freetype \
+  mingw64/mingw-w64-x86_64-fontconfig \
+  mingw64/mingw-w64-x86_64-librsvg \
+  mingw64/mingw-w64-x86_64-libxml2
```
@chearon I had to add this libxml2 line. Do I also need to add something to line 3? (`deps=`)

chearon replied:
Yeah you do to get the `.lib` files. Did it work without doing that?

zbjornson replied:
Yeah, somehow that lib got in there. Output of the `tar -czvf` command from the logs in my last post (last line):

```
2020-05-01T04:06:38.2807995Z Release/
2020-05-01T04:06:38.2808850Z Release/libgdk_pixbuf-2.0.so.0
2020-05-01T04:06:38.3231600Z Release/libgthread-2.0.so.0
2020-05-01T04:06:38.3237935Z Release/libpango-1.0.so.0
2020-05-01T04:06:38.4053897Z Release/libz.so.1
2020-05-01T04:06:38.4099015Z Release/libgobject-2.0.so.0
2020-05-01T04:06:38.5040638Z Release/libharfbuzz.so.0
2020-05-01T04:06:38.8123409Z Release/libstdc++.so.6
2020-05-01T04:06:38.8662366Z Release/libgif.so.7
2020-05-01T04:06:38.8723402Z Release/libfontconfig.so.1
2020-05-01T04:06:38.9515542Z Release/libgmodule-2.0.so.0
2020-05-01T04:06:38.9525404Z Release/libpcre.so.1
2020-05-01T04:06:38.9962821Z Release/librsvg-2.so.2
2020-05-01T04:06:41.1066899Z Release/libpangocairo-1.0.so.0
2020-05-01T04:06:41.1218360Z Release/libexpat.so.1
2020-05-01T04:06:41.1531592Z Release/libgio-2.0.so.0
2020-05-01T04:06:41.7131079Z Release/.deps/
2020-05-01T04:06:41.7134142Z Release/.deps/Release/
2020-05-01T04:06:41.7136499Z Release/.deps/Release/canvas.node.d
2020-05-01T04:06:41.7139171Z Release/.deps/Release/obj.target/
2020-05-01T04:06:41.7141421Z Release/.deps/Release/obj.target/canvas.node.d
2020-05-01T04:06:41.7143937Z Release/.deps/Release/obj.target/canvas/
2020-05-01T04:06:41.7146407Z Release/.deps/Release/obj.target/canvas/src/
2020-05-01T04:06:41.7148812Z Release/.deps/Release/obj.target/canvas/src/bmp/
2020-05-01T04:06:41.7151117Z Release/.deps/Release/obj.target/canvas/src/bmp/BMPParser.o.d
2020-05-01T04:06:41.7153620Z Release/.deps/Release/obj.target/canvas/src/CanvasPattern.o.d
2020-05-01T04:06:41.7161364Z Release/.deps/Release/obj.target/canvas/src/Backends.o.d
2020-05-01T04:06:41.7165449Z Release/.deps/Release/obj.target/canvas/src/color.o.d
2020-05-01T04:06:41.7169159Z Release/.deps/Release/obj.target/canvas/src/init.o.d
2020-05-01T04:06:41.7181367Z Release/.deps/Release/obj.target/canvas/src/closure.o.d
2020-05-01T04:06:41.7185210Z Release/.deps/Release/obj.target/canvas/src/Image.o.d
2020-05-01T04:06:41.7189211Z Release/.deps/Release/obj.target/canvas/src/Canvas.o.d
2020-05-01T04:06:41.7192989Z Release/.deps/Release/obj.target/canvas/src/CanvasGradient.o.d
2020-05-01T04:06:41.7196962Z Release/.deps/Release/obj.target/canvas/src/backend/
2020-05-01T04:06:41.7200641Z Release/.deps/Release/obj.target/canvas/src/backend/ImageBackend.o.d
2020-05-01T04:06:41.7204171Z Release/.deps/Release/obj.target/canvas/src/backend/Backend.o.d
2020-05-01T04:06:41.7205217Z Release/.deps/Release/obj.target/canvas/src/backend/PdfBackend.o.d
2020-05-01T04:06:41.7208066Z Release/.deps/Release/obj.target/canvas/src/backend/SvgBackend.o.d
2020-05-01T04:06:41.7210907Z Release/.deps/Release/obj.target/canvas/src/register_font.o.d
2020-05-01T04:06:41.7213821Z Release/.deps/Release/obj.target/canvas/src/CanvasRenderingContext2d.o.d
2020-05-01T04:06:41.7216743Z Release/.deps/Release/obj.target/canvas/src/ImageData.o.d
2020-05-01T04:06:41.7219762Z Release/canvas.node
2020-05-01T04:06:41.7443529Z Release/libpixman-1.so.0
2020-05-01T04:06:42.0423568Z Release/libfreetype.so.6
2020-05-01T04:06:42.2641755Z Release/libcroco-0.6.so.3
2020-05-01T04:06:42.3323068Z Release/libcairo.so.2
2020-05-01T04:06:42.6984065Z Release/libfribidi.so.0
2020-05-01T04:06:42.7072388Z Release/libjpeg.so.62
2020-05-01T04:06:42.7332160Z Release/libpng16.so.16
2020-05-01T04:06:42.7949087Z Release/libglib-2.0.so.0
2020-05-01T04:06:43.0451484Z Release/libpangoft2-1.0.so.0
2020-05-01T04:06:43.0756996Z Release/libffi.so.6
2020-05-01T04:06:43.0767866Z Release/obj.target/
2020-05-01T04:06:43.0770916Z Release/obj.target/canvas.node
2020-05-01T04:06:43.1032178Z Release/obj.target/canvas/
2020-05-01T04:06:43.1038588Z Release/obj.target/canvas/src/
2020-05-01T04:06:43.1045185Z Release/obj.target/canvas/src/bmp/
2020-05-01T04:06:43.1050692Z Release/obj.target/canvas/src/bmp/BMPParser.o
2020-05-01T04:06:43.1056642Z Release/obj.target/canvas/src/closure.o
2020-05-01T04:06:43.1061819Z Release/obj.target/canvas/src/CanvasRenderingContext2d.o
2020-05-01T04:06:43.1166693Z Release/obj.target/canvas/src/ImageData.o
2020-05-01T04:06:43.1176701Z Release/obj.target/canvas/src/register_font.o
2020-05-01T04:06:43.1180824Z Release/obj.target/canvas/src/Image.o
2020-05-01T04:06:43.1207499Z Release/obj.target/canvas/src/CanvasPattern.o
2020-05-01T04:06:43.1225973Z Release/obj.target/canvas/src/backend/
2020-05-01T04:06:43.1229154Z Release/obj.target/canvas/src/backend/ImageBackend.o
2020-05-01T04:06:43.1232328Z Release/obj.target/canvas/src/backend/SvgBackend.o
2020-05-01T04:06:43.1241894Z Release/obj.target/canvas/src/backend/Backend.o
2020-05-01T04:06:43.1256586Z Release/obj.target/canvas/src/backend/PdfBackend.o
2020-05-01T04:06:43.1260209Z Release/obj.target/canvas/src/color.o
2020-05-01T04:06:43.1284377Z Release/obj.target/canvas/src/init.o
2020-05-01T04:06:43.1287528Z Release/obj.target/canvas/src/Backends.o
2020-05-01T04:06:43.1290889Z Release/obj.target/canvas/src/CanvasGradient.o
2020-05-01T04:06:43.1299270Z Release/obj.target/canvas/src/Canvas.o
2020-05-01T04:06:43.1369933Z Release/libxml2.so.2
```

chearon replied:
Oh duh, it's been a while since I've looked at this. You don't need the `.lib` file unless you're linking with a library. `libxml2` is included because _other_ libraries link to it. So it should actually _not_ be put in `deps=`.

zbjornson replied:
I posted the Ubuntu log instead of the Windows log. But yeah, looking at the Windows log, what you just said sounds correct!

zbjornson commented:
3rd time's a charm.

https://github.com/zbjornson/node-canvas/actions/runs/92779780
https://github.com/zbjornson/node-canvas/releases/tag/untagged-f98836e32200015876d6

chearon commented on prebuild/macOS/.gitignore:
```diff
@@ -0,0 +1 @@
+macdylibbundler/
```
This was vestigial, should be able to remove this since it uses `macpack` now

zbjornson replied:
Removed.

chearon commented on .github/workflows/prebuild.yaml:
```diff
@@ -0,0 +1,234 @@
+          . prebuild/$OS/bundle.sh
+
+      - name: Test binary
+        run: node -e "require('.')"
```
IIRC I couldn't run the tests here because they weren't in the NPM tarball - that or I was just lazy :D. Might be good opportunity to do the tests instead of just testing loading. Possibly uninstall Cairo et all first to make sure the prebuilds did work.

chearon commented on .github/workflows/prebuild.yaml:
```diff
@@ -0,0 +1,234 @@
+        run: |
+          npm install -g node-gyp
+          npm install --ignore-scripts
+          . prebuild/$OS/preinstall.sh
```
Scripts aren't using `$OS` so you could just change these to be `prebuild/${{runner.os}}` and delete the OS variable

chearon commented on .github/workflows/prebuild.yaml:
```diff
@@ -0,0 +1,234 @@
+            const releases = await github.repos.listReleases({
+              owner: process.env.GITHUB_REPOSITORY.split("/")[0],
+              repo: process.env.GITHUB_REPOSITORY.split("/")[1]
+            });
```
Could save 3 lines here (* 3 OSes)

```javascript
            const [owner, repo] = process.env.GITHUB_REPOSITORY.split("/");
            const releases = await github.repos.listReleases({owner, repo});
```

chearon commented on .github/workflows/prebuild.yaml:
```diff
@@ -0,0 +1,234 @@
+      - name: Setup
+        run: |
+          msys2do npm install -g node-gyp
+          msys2do npm install --ignore-scripts
```
Technically the `npm`/`node-gyp` lines shouldn't need `msys2do` should they?

chearon commented:
One thing to think about: say we merge this and release prebuilds for node 14, but people ask for 2.6.0 node 14 prebuilds, and 2.5.0, and so on... we will have to keep amending the file to trigger those builds, then put it back to 2.6.1 again, which would be a lot of force-pushes to `master`.

It looks like you [might be able to trigger events manually](https://help.github.com/en/actions/reference/events-that-trigger-workflows#triggering-new-workflows-using-a-personal-access-token) using a personal access token. There's a ton of other events on that page we might consider.

domoritz commented:
Another option would be to also allow builds from branches, not just master. You could also allow builds from branches of the form `v*` and then have branches for each version (e.g. `v2.6.1`).

chearon commented:
I was thinking about that too, but the downside is we'd have a bunch of branches with commits that aren't really useful after the build happens.

zbjornson commented:
I can make the `strategy.matrix` define a set of tags instead of a single tag, so it could do multiple versions in one push. When a new version of Node.js comes out, we'd use one push for that new Node.js version and some array canvas releases. When a new version of canvas is released, we'd use one push for an array of Node.js and one version of canvas. Would that work for you guys?

As far as triggers, another option is a single "deploy trigger" branch where we just make commits to trigger deploys.

> say we merge this and release prebuilds for node 14, but people ask for 2.6.0 node 14 prebuilds, and 2.5.0, and so on...

Another issue ... 2.6.0 as published to NPM uses node-gfx/node-canvas-prebuilt for its tarballs, so we need to manually upload artifacts there, or temporarily tinker with this GitHub Action to publish the artifacts there instead of here, or say "Node 14 requires 2.6.1 or later." NAPI would alleviate this.

zbjornson commented:
Summary of changes in last force-push:
- Added instructions for triggering builds.
- Added `jobs.*.strategy.matrix.canvas_tag` to allow building multiple version of canvas at once. (Use when new versions of Node.js are released.)
- Renamed `CANVAS_PREBUILT_VERSION` to `UPLOAD_TO`, and made it optional (defaulting to the same tag name of the release being built).
- Changed the two version environment variables to require the "v" prefix instead of adding it automatically. Since these are GitHub tag names, this is a bit more flexible if e.g. `UPLOAD_TO` should be `testing-123`.
- Made full test suite run instead of just `require(".")`, after uninstalling the dependencies.
- Made the CI tests not run when only `prebuild.yaml` changes.
- Addressed @chearon's comments.
- Shortened the upload script.

As far as strategy for dealing with releases that were published pointing to node-gfx/node-canvas-prebuilt, I don't mind manually copying the build artifacts for the last several releases.

chearon commented:
Manually copying to node-gfx is definitely the way to go.

The tag matrix is a good idea except I think it will still be noisy (you have to reset the version array to 1 element after building them all). Were you thinking of still having a trigger branch? One thing that's very important to remember is that the `gyp` files in the CI folders will change, so we will need a new trigger branch each time they change.

There is one other big-ish thing I should comment on: the build matrix normally iterates the operating system. That has obvious deduplication upsides. It doesn't seem impossible to do, but there might be some potholes along the way. Thoughts?

zbjornson commented:
Some more observations after trying to make prebuilds for Node.js v14 and older Canvas versions:

* So that the prebuild scripts were available in the checkout of older releases, I added

      git fetch --prune --unshallow
      git checkout origin/zb/wf-release -- prebuild

  to the workflow file. (@chearon this is one approach to dealing with what you said about the `gyp` files changing, but also, now that the prebuild `gyp` files will be versioned with the canvas code, I don't think this will be an issue.)

  That worked, except on Linux where I hit https://github.com/actions/checkout/issues/126 (workaround exists in that thread).

* Building some old canvas versions is impossible because the tests fail, now that this runs the full suite. (Trying to build v2.6.0 I hit the failure that was fixed by 4b362b4f4a0db0ffe7ba0bbbef2bbf1218465be0.)

For now I think we should keep prebuild triggering isolated in a separate branch until we get a rhythm down, and can move toward automatic triggers when releases are made. I'm inclined to not release prebuilds for older versions of canvas, but we could allow the tests to fail if we decide to do it.

> the build matrix normally iterates the operating system

Because the Linux build uses a container, this isn't possible. macOS and Windows could run together (that's why I previously had `${{ runner.OS }}` in some places) but they need a bunch of `if` clauses because Windows needs `msys2do`. I'm more hopeful that GitHub will support YAML anchors (https://github.community/t5/GitHub-Actions/Support-for-YAML-anchors/td-p/30336) soon so that e.g. the upload script doesn't have to be repeated.

---

Summary proposal:
* Merge this into master, but set `on.push.branches` to `"build-trigger"`
* When the next release of canvas is made, we rebase the build-trigger branch on master, set `matrix.canvas_tag` to that tag name, `matrix.node` to major versions 8-14, push and things should just work.

Thoughts?

LinusU commented:
> Summary proposal:
> 
> * Merge this into master, but set `on.push.branches` to `"build-trigger"`
> * When the next release of canvas is made, we rebase the build-trigger branch on master, set `matrix.canvas_tag` to that tag name, `matrix.node` to major versions 8-14, push and things should just work.
> 
> Thoughts?

Sounds good to me 👍 

If we see that this works reliably I guess that we could do `on: push: tags: - v*`, and then use `${GITHUB_REF/refs\/tags\/v/}` to determine which version to build?

chearon commented:
@LinusU yeah, I believe that is exactly the plan.

Once we get this merged I'll probably bring over the Dockerfile, then move some of the PRs/issues from node-canvas-prebuilt to here. Some work has been done towards MUSL (node-gfx/node-canvas-prebuilt#116) and maybe someone can do ARM because I don't think I'm going to. I'm hoping that having everything centralized will make it easier for others to contribute!

chearon commented:
Looks like we're all in agreement, so merging! Huge thanks again to @zbjornson!
